Mathlib/AlgebraicGeometry/AffineScheme.lean

```diff
@@ -895,36 +917,136 @@ theorem of_affine_open_cover {X : Scheme} {P : X.affineOpens → Prop}
+  Scheme.isoSpec_image_zeroLocus
+
 /-- If `X` is an affine scheme, every closed set of `X` is the zero locus
 of a set of global sections. -/
-lemma Scheme.eq_zeroLocus_of_isClosed_of_isAffine (X : Scheme.{u}) [IsAffine X] (s : Set X) :
+lemma eq_zeroLocus_of_isClosed_of_isAffine [IsAffine X] (s : Set X) :
     IsClosed s ↔ ∃ I : Ideal (Γ(X, ⊤)), s = X.zeroLocus (I : Set Γ(X, ⊤)) := by
   refine ⟨fun hs ↦ ?_, ?_⟩
   · let Z : Set (Spec <| Γ(X, ⊤)) := X.toΓSpecFun '' s
     have hZ : IsClosed Z := (X.isoSpec.hom.homeomorph).isClosedMap _ hs
     obtain ⟨I, (hI : Z = _)⟩ := (PrimeSpectrum.isClosed_iff_zeroLocus_ideal _).mp hZ
     use I
-    simp only [← Scheme.toΓSpec_preimage_zeroLocus_eq, ← hI, Z]
+    simp only [← Scheme.toSpecΓ_preimage_zeroLocus, ← hI, Z]
     erw [Set.preimage_image_eq _ (bijective_of_isIso X.isoSpec.hom.base).injective]
   · rintro ⟨I, rfl⟩
     exact zeroLocus_isClosed X I.carrier
 
+open Set.Notation in
+lemma Opens.toSpecΓ_preimage_zeroLocus {X : Scheme.{u}} (U : X.Opens)
+    (s : Set Γ(X, U)) :
+    U.toSpecΓ.base ⁻¹' PrimeSpectrum.zeroLocus s = U.1 ↓∩ X.zeroLocus s := by
+  rw [toSpecΓ, Scheme.comp_base, TopCat.coe_comp, Set.preimage_comp, Spec.map_base]
+  erw [PrimeSpectrum.preimage_comap_zeroLocus]
+  rw [Scheme.toSpecΓ_preimage_zeroLocus]
+  show _ = U.ι.base ⁻¹' (X.zeroLocus s)
+  rw [Scheme.preimage_zeroLocus, U.ι_app_self, ← zeroLocus_map_of_eq _ U.ι_preimage_self,
+    ← Set.image_comp, ← RingHom.coe_comp, ← CommRingCat.hom_comp]
+  congr!
+  simp [← Functor.map_comp]
+  rfl
+
+end Scheme
+
+lemma IsAffineOpen.fromSpec_preimage_zeroLocus {X : Scheme.{u}} {U : X.Opens}
+    (hU : IsAffineOpen U) (s : Set Γ(X, U)) :
+    hU.fromSpec.base ⁻¹' X.zeroLocus s = PrimeSpectrum.zeroLocus s := by
+  ext x
+  suffices (∀ f ∈ s, ¬¬ f ∈ x.asIdeal) ↔ s ⊆ x.asIdeal by
+    simpa [← hU.fromSpec_image_basicOpen, -not_not] using this
+  simp_rw [not_not]
+  rfl
+
+lemma IsAffineOpen.fromSpec_image_zeroLocus {X : Scheme.{u}} {U : X.Opens}
+    (hU : IsAffineOpen U) (s : Set Γ(X, U)) :
+    hU.fromSpec.base '' PrimeSpectrum.zeroLocus s = X.zeroLocus s ∩ U := by
+  rw [← hU.fromSpec_preimage_zeroLocus, Set.image_preimage_eq_inter_range, range_fromSpec]
+
+open Set.Notation in
+lemma Scheme.zeroLocus_inf (X : Scheme.{u}) {U : X.Opens} (I J : Ideal Γ(X, U)) :
+    X.zeroLocus (U := U) ↑(I ⊓ J) = X.zeroLocus (U := U) I ∪ X.zeroLocus (U := U) J := by
+  suffices U.1 ↓∩ (X.zeroLocus (U := U) ↑(I ⊓ J)) =
+      U.1 ↓∩ (X.zeroLocus (U := U) I ∪ X.zeroLocus (U := U) J) by
+    ext x
+    by_cases hxU : x ∈ U
+    · simpa [hxU] using congr(⟨x, hxU⟩ ∈ $this)
+    · simp only [Submodule.inf_coe, Set.mem_union,
+        codisjoint_iff_compl_le_left.mp (X.codisjoint_zeroLocus (U := U) (I ∩ J)) hxU,
+        codisjoint_iff_compl_le_left.mp (X.codisjoint_zeroLocus (U := U) I) hxU, true_or]
+  simp only [← U.toSpecΓ_preimage_zeroLocus, PrimeSpectrum.zeroLocus_inf I J,
+    Set.preimage_union]
+
+lemma Scheme.zeroLocus_biInf
```

Comment:
What is the explanation behind the name "**b**iInf".

Reply:
It means bounded-iInf, referring to the `∈ s` part.